v2/cmd/save.go

```diff
@@ -0,0 +1,259 @@
+	"github.com/otiai10/copy"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"k8s.io/klog/v2"
+)
+
+// saveCmd represents the save command
+var saveCmd = &cobra.Command{
+	Use:   "save",
+	Short: "Save full license text locally",
+	Long:  `Save full license text and source code locally to be compliant.`,
+	Run: func(cmd *cobra.Command, args []string) {
+		config, err := config.Load("")
+		if err != nil {
+			klog.ErrorS(err, "Failed: load config")
+			os.Exit(1)
+		}
+		info, err := loadInfo(config.Module.Csv.Path)
+		if err != nil {
+			klog.ErrorS(err, "Failed: load license info csv")
+			os.Exit(1)
+		}
+		err = complyWithLicenses(info, *config)
+		if err != nil {
+			klog.ErrorS(err, "Failed: comply with licenses")
+			os.Exit(1)
+		}
+		klog.Flush()
+	},
+}
+
+func init() {
+	rootCmd.AddCommand(saveCmd)
+
+	// Here you will define your flags and configuration settings.
+
+	// Cobra supports Persistent Flags which will work for this command
+	// and all subcommands, e.g.:
+	// saveCmd.PersistentFlags().String("foo", "", "A help for foo")
+
+	// Cobra supports local flags which will only run when this command
+	// is called directly, e.g.:
+	// saveCmd.Flags().BoolP("toggle", "t", false, "Help message for toggle")
+}
+
+const defaultNoticesPath = "NOTICES"
+const defaultLicenseSubPath = "licenses.txt"
+const defaultSrcPath = "src"
+
+// Dir permission needs execute bit for `cd` or `ls` commands
+// ref: https://www.tutorialspoint.com/unix/unix-file-permission.htm
+const permDirCurrentUser = 0700
+const permFileCurrentUser = 0600
+
+// The following code is largely inspired by
+// https://github.com/google/go-licenses/blob/master/save.go, which is licensed under:
+//
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// license compliance requirement type
+type ComplianceReq string
+
+const (
+	// We do not allow unknown licenses.
+	Unknown ComplianceReq = "Unknown"
+	// We need to redistribute the entire source directory to be compliant,
+	// example licenses: GPL, MPL, etc.
+	RedistributeSource ComplianceReq = "DistributeSource"
+	// We need to redistribute full text license and a copyright notice to be
+	// compliant: most other licenses.
+	RedistributeNotice ComplianceReq = "DistributeNotice"
+)
+
+// Determines compliance requirement type of a license, returns ComplianceReq.
+// license can be a list of licenses like "Apache-2.0 / MIT", this method returns
```

Comment:
Many things in the Rust ecosystem are dual licensed so you can use it with the Apache-2.0 license or the MIT license

Reply:
Good to know! Do you know of any examples I can look at?

I also just learned that there's SPDX license operators 😵  (e.g. AND, OR) https://spdx.github.io/spdx-spec/appendix-IV-SPDX-license-expressions/. Probably don't need to tackle this in this PR, but something to be aware of.

Reply:
https://github.com/rust-lang/rust :wink: 

Some more:
https://github.com/rust-lang/rustup
https://github.com/rust-embedded/cross
https://github.com/cloudflare/wrangler
https://github.com/sharkdp/bat
https://github.com/XAMPPRocky/tokei

Some other variations:
APL2.0 and/or (not sure which) UPL: https://github.com/oracle/railcar
MIT or unlicense: https://github.com/boyter/scc
Part LGPL2 part GPL2: https://github.com/libfuse/libfuse
most things in KDE are gpl2 lgpl21 fdl12 or some other strange combos

Reply:
Thanks @06kellyjac!

I also got some examples, but looks like you gave enough.

My main rationale was that the portion of these repos is relatively low, that doesn't make it worth introducing an array type until now, because config syntax will be more verbose.